### PR TITLE
Miscellaneous improvements to the mysql legacy relation

### PR DESCRIPTION
**Pull request by shayancanonical**
## Issue
The following are discrepancies between the legacy mysql relation code in vms vs. k8s:
- We are not correctly handling exception cases when calling `get_cluster_primary_address` (where the return value is None)
- We have no exception handling when calling `does_mysql_user_exist`
- We are accessing the `root-password` directly from the app peer databag rather than using the `get_secret` helper (which will also retrieve it from the databag in juju v2.9.43)
- We are not deleting legacy mysql relation data from the peer databag upon relation broken

## Solution
- Check if the return value for `get_cluster_primary_address` is None and if so, avoid calling `.split()` on None
- Add exception handling around `does_mysql_user_exist`
- Access `root-password` using the `get_secret` helper
- Delete the relation data from the peer databag upon relation broken

### CODE REVIEW

File: src/relations/mysql.py [hunk 1]
```diff
@@ -181,10 +197,10 @@ def _on_mysql_relation_created(self, event: RelationCreatedEvent) -> None:
 
         updates = {
             "database": database,
-            "host": primary_address,
+            "host": primary_address.split(":")[0],
             "password": password,
             "port": "3306",
-            "root_password": self.charm.app_peer_data["root-password"],
+            "root_password": self.charm.unit.get_secret("app", "root-password"),
```
[the remark this answers is not in the document]
Comment (shayancanonical) [reply src/relations/mysql.py]: indeed it is not. changed in [cdefc8e](https://github.com/canonical/mysql-operator/pull/254/commits/cdefc8e22f2a74e2387e001f8364527f0d9c125c)

File: src/relations/mysql.py [hunk 2]
```diff
@@ -170,7 +186,7 @@ def _on_mysql_relation_created(self, event: RelationCreatedEvent) -> None:
                 unit_name="mysql-legacy-relation",
             )
 
-            primary_address = self.charm._mysql.get_cluster_primary_address().split(":")[0]
+            primary_address = self.charm._mysql.get_cluster_primary_address()
```
[the remark this answers is not in the document]
Comment (shayancanonical) [reply src/relations/mysql.py:191]: yes, good call. checked in a730e8f5950fff679145314288bfdd79c327f79e

**Review by taurus-forever — APPROVED** [review]
This is great, thank you!
P.S. I would use error state instead of blocked, but we will improve this in a general way for all blocked states we use.

**Review by paulomach — APPROVED** [review]
lgtm

**Review by paulomach — DISMISSED** [review]

**Review by taurus-forever — DISMISSED** [review]
Nice findings! Tnx!

**Review by paulomach — APPROVED** [review]

File: src/relations/mysql.py [hunk 3]
```diff
@@ -180,9 +198,13 @@ def _on_mysql_relation_created(self, event: RelationCreatedEvent) -> None:
...
+        if not primary_address:
+            self.charm.unit.status = BlockedStatus("Failed to retrieve cluster primary address")
+            return
+
         updates = {
             "database": database,
-            "host": primary_address,
+            "host": primary_address.split(":")[0],
             "password": password,
             "port": "3306",
```
[the remark this answers is not in the document]
Comment (shayancanonical) [reply src/relations/mysql.py:209]: We are hardcoding `3306` as the port in numerous places - I'd say we should address it holistically as a separate effort

**Review by paulomach — APPROVED** [review]

**Pull request merged**